File: xxGreetings_wLogger/Greetings_wLogger.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <array>
#include "spdlog/spdlog.h"
#include "spdlog/fmt/ranges.h"
#include "spdlog/sinks/basic_file_sink.h"
#include "Userinfo.h"

void ShowGreetings(const std::string& a_filename)
{
	std::ifstream file(a_filename);
	while (file.good())
	{
		std::string line;
		std::getline(file, line);
		if (!line.empty())
		{
			std::cout << "* " << line << "\n";
		}
	}
}

bool ContainsShouting(const std::string& a_greeting)
{
	size_t contigousUpperCase = 0;
	for (char c : a_greeting)
	{
		contigousUpperCase = isupper(c) ? contigousUpperCase + 1 : 0;
		if (3 == contigousUpperCase)
		{
			return true;
		}
	}

	return false;
}

static constexpr const char *FILENAME = "greetings.txt";

static void SetupLogToFile(const std::string& a_logfilePath)
{
	auto fileLogger = spdlog::basic_logger_mt("main", a_logfilePath.c_str(), true);
    spdlog::set_default_logger(fileLogger);
    spdlog::flush_every(std::chrono::seconds(1));
}

int main()
{
	SetupLogToFile("greetings.log");

    std::string username = GetDomainName() + '\\' + GetUsername();

    spdlog::debug("This will be a DEBUGGING log entry");
    spdlog::info("This will be an INFORMATIONAL log entry");
    spdlog::warn("This will be a WARNING log entry");
    spdlog::error("This will be an ERROR log entry");
    spdlog::critical("This will be a CRITICAL log entry");

	std::cout << "Hello, do you have a greeting for your coworkers? (Press enter to just see others' greetings)\n";
	std::cout << ">>> ";
	std::string line;
	std::getline(std::cin, line);

	if (ContainsShouting(line))
	{
		std::cout << "Please use moderate tones...\n";
		return 1;
	}

	std::cout << "Here are some greetings from others...\n";
	ShowGreetings(FILENAME);

	if (!line.empty())
	{
		std::ofstream out(FILENAME, std::ios_base::app);
		out << line << "\n";
	}

	return 0;
}

File: SharpGreetings_wLogger/Greetings_wLogger.cs
// Required nuget packages:
// Microsoft.Extensions.Logging
// Microsoft.Extensions.Logging.Debug

using Microsoft.Extensions.Logging;

using System;
using System.IO;

namespace SharpGreetings
{
    class Greetings
    {
        static ILoggerFactory s_loggerFactory = LoggerFactory.Create(builder =>
        {
            builder
                .AddFilter("Microsoft", LogLevel.Warning)
                .AddFilter("System", LogLevel.Warning)
                .AddFilter("SharpGreetings.Greetings", LogLevel.Debug)
                .AddDebug();
        });

        private static ILogger<T> CreateLogger<T>()
        {
            ILogger<T> logger = s_loggerFactory.CreateLogger<T>();
            return logger;
        }

        static ILogger<Greetings> theLogger = CreateLogger<Greetings>();

        public static void Main(string[] args)
        {
            string username = Environment.UserDomainName + '\\' + Environment.UserName;

            theLogger.LogDebug("This will be an DEBUGGING log entry");
            theLogger.LogInformation("This will be an INFORMATIONAL log entry");
            theLogger.LogWarning("This will be an WARNING log entry");
            theLogger.LogError("This will be an ERROR log entry");
            theLogger.LogCritical("This will be an CRITICAL log entry");

            Console.Write("Hello, do you have a greeting for your coworkers? (Press enter to just see others' greetings)\n");
            Console.Write(">>> ");
            string newLine = Console.ReadLine();

            if (ContainsShouting(newLine))
            {
                Console.Write("Please use moderate tones...\n");
                return;
            }

            Console.Write("Here are some greetings from others...\n");
            ShowGreetings(FILENAME);

            if (newLine.Length != 0)
            {
                File.AppendAllText(FILENAME, newLine + "\n");
            }
        }

        private static bool ContainsShouting(string a_greeting)
        {
            uint contigousUpperCase = 0;
            foreach (char c in a_greeting)
            {
                contigousUpperCase = char.IsUpper(c) ? contigousUpperCase + 1 : 0;
                if (3 == contigousUpperCase)
                {
                    return true;
                }
            }

            return false;
        }

        private static readonly string FILENAME = "greetings.txt";

        private static void ShowGreetings(string a_filename)
        {
            try
            {
                string[] lines = File.ReadAllLines(a_filename);
                foreach (string line in lines)
                {
                    Console.Write("* ");
                    Console.WriteLine(line);
                }
            }
            catch (FileNotFoundException)
            {
                // ignore
            }
        }
    }
}
